programs/gamma/src/utils/swap_referral.rs

```diff
@@ -0,0 +1,62 @@
+use anchor_lang::prelude::*;
+use referral::ReferralAccount;
+use referral::REFERRAL_ATA_SEED;
+use spl_token::state::{Account as SplTokenAccount, GenericTokenAccount};
+
+pub struct ReferralDetails<'c, 'info> {
+    pub share_bps: u16,
+    pub referral_token_account: &'c AccountInfo<'info>,
+}
+
+pub fn extract_referral_info<'c, 'info>(
+    input_token_mint: Pubkey,
+    check_referral_project_key: Pubkey,
+    remaining_accounts: &'c [AccountInfo<'info>],
+) -> Result<Option<ReferralDetails<'c, 'info>>> {
+    // We take exactly three accounts:
+    // 1. The project account
+    // 2. The referral account
+    // 3. The referral token-account
```

Comment:
We don't need all 3 accounts, we only need referral token-account.

In the logic below we are using the project account's public key only which we can get from the amm_config.